### ISSUE TITLE
CCing notifications to collaborators

### ISSUE THREAD

**Issue by torohill** [osTicket/osTicket-1.8#1764]
Currently when a staff member posts a response to a ticket that has collaborators, the submitter and each collaborator gets an individual email notification (ie. an email 'To' them).

In the case where the collaborators were added via CC on the original ticket email, we have found that this can cause some confusion with the submitter/collaborators, namely:

1. The email notification contains no indication that it has been sent to all the collaborators. This has caused people to ask us if the other collaborators received a notification or even to forward the notification to the other collaborators.
2. If someone replies to the email notification and again CCs the collaborators (just to make sure they see the reply) and the `Autoresponder > New Message > Participants: Send new activity notice` setting is turned on, then collaborators will receive both the CCed reply and another email notification from osTicket.
3. If someone replies to the email notification (and doesn't CC the collaborators because they assume osTicket will handle it) and the `Autoresponder > New Message > Participants: Send new activity notice` is turned off then the collaborators never see the response.

I can see a couple of possible solutions to this issue: 

1. In the email notification template add a note that the collaborators have also been notified, and preferable include a list of the collaborators names. I'm not sure if it is currently possible to include a list of collaborators names in an email template.
2. Instead of sending email notifications to each collaborator individually, send a single email notification 'To' the submitter and CC it to the collaborators. 

I think the 2nd solution is preferable as it addresses the 3 points from above:

1. The collaborators are CCed so it easy to tell when viewing the email notification that they have been included.
2. Most people will 'reply all' to a notification, which will include the collaborators, hence there is no need to CC them again. There probably would also need to be a setting to turn on CCing of email notifications to collaborators, and this setting would need to be mutually exclusive with the `Autoresponder > New Message > Participants: Send new activity notice` setting, to ensure that people don't receive the CCed reply and another email notification from osTicket.
3. Again, most people will 'reply all' to make sure that collaborators are included.

In general, the 2nd solution allows people to continue to use their email client and CC collaborators in the same way they do for normal emails. It also give them control over exactly who is CCed/notified on a reply.

I am happy to implement the 2nd solution and submit it as a pull request if you think it is worthwhile. Do you see any downsides to this approach? Can you think of any other possible solutions that would be preferable?

Thanks

**Comment by greezybacon** [comment]
So, in short, neither solution will work. However, I agree that the original point number two is quite valid. The system should not send the notification email if the recipients were in the `Cc:` header. We should fix that.

The recipients cannot appear in the Cc header (and receive a single email sent to all the recipients), because we have the capability to include individual auto-login tokens with `%{recipient.ticket_link}`. Therefore, sending an email to a party would mean abandoning the auto-login link allowing users one-click access to the web interface to continue the conversation.

The collaborator listing can be included in the response templates with `%{ticket.recipients}`

**Comment by mfelber** [comment]
@greezybacon Thank you so much for the hint with the variable %{ticket.recipients}. This makes many people happy around me :-) I have never seen this variable before (forum, github, supported variables osticket, ntozier's site http://tmib.net), is this a new variable? Maybe you should list it also in the supported variable section in osticket, I think many other people will benefit too from this "stealth" variable ;-)

**Comment by mfelber** [comment]
Now, because I'm aware that this variable exists, I found one entry in the forum that mentioned this variable before (May 2014, so I think it is not really new) when I search with this term: "ticket.recipients".

**Comment by ntozier** [comment]
I've added this to my list at:
http://tmib.net/osticket-191-email-template-variables

Sorry I must have missed it.

**Comment by mfelber** [comment]
@ntozier No need to say sorry. I think we are all really happy that somebody like you support us/the project so helpful :-) Thank you for adding this to the list!

**Comment by JanKraljic** [comment]
Beside list of %{ticket.recipients} we would like also their email address something like %{ticket.recipients.email}. So that at can all  Collaborators with email can be written in message. Thanks for tip!

**Comment by greezybacon** [comment]
@JanKraljic thanks. For now that would be a separate enhancement request

**Comment by greezybacon** [comment]
@torohill could you try the patch in #1871 to see if it works to eliminate duplicate mail to collaborators?

**Comment by torohill** [comment]
@greezybacon Sorry I didn't reply get a chance to test this before now (it's been a busy couple of months). The patch in #1871 doesn't fix this problem for me, based on my testing. Below is a diff this does work for me. Also, I think it's probably a good idea to run both emails through `mb_strtolower` before comparing them to make it case insensitive.

Shall I go ahead and submit this fix as a pull request?

```diff
diff --git a/include/class.ticket.php b/include/class.ticket.php
index 6b54bb0..0d30111 100644
--- a/include/class.ticket.php
+++ b/include/class.ticket.php
@@ -1117,7 +1117,7 @@ class Ticket {
             // Skip all the other recipients of the message
             foreach ($entry->getAllEmailRecipients() as $R) {
                 foreach ($recipients as $R2) {
-                    if ($R2->getEmail() == ($R->mailbox.'@'.$R->hostname)) {
+                    if ($R2->getEmail() == ($R->mailbox.'@'.$R->host)) {
                         $skip[$R2->getUserId()] = true;
                         break;
                     }
```

**Comment by greezybacon** [comment]
@torohill thanks for the bug report. I'll fix it

**Comment by greezybacon** [comment]
It's fixed in 2484447
